[REEF-1257] Add TaskCloseHandler in TaskRunTime

jwang98052 opened the pull request:
*Inject TaskCloseHandler  in TaskRunTime
*Handle Close event
*Add DefaultCloseTaskHandler
*Updated test cases

JIRA: [REEF-1257](https://issues.apache.org/jira/browse/REEF-1257)

This closes #

markusweimer commented:
looking at it.

markusweimer commented on lang/cs/Org.Apache.REEF.Common/Runtime/Evaluator/Task/TaskRuntime.cs:
```diff
@@ -159,6 +163,11 @@ public bool HasEnded()
             Logger.Log(Level.Info, string.Format(CultureInfo.InvariantCulture, "Trying to close Task {0}", TaskId));
+            if (message != null)
+            {
+                Logger.Log(Level.Info, string.Format(CultureInfo.InvariantCulture, "Message received from task close event {0}.", Encoding.UTF8.GetString(message)));
```
This shouldn't go to the `Info` level but instead one of the more verbose ones.

jwang98052 replied:
Done

markusweimer commented on lang/cs/Org.Apache.REEF.Common/Tasks/Defaults/DefaultCloseTaskHandler.cs:
```diff
@@ -0,0 +1,50 @@
+    /// of a task. Throws an exception by default, since a task should not have received
+    /// a suspension event if the handler is not bound explicitly.
+    /// </summary>
+    internal sealed class DefaultCloseTaskHandler : IObserver<ICloseEvent>
```
Please rename to `DefaultTaskCloseHandler`. That way, auto-completion will go from the least specific to the most specific type.

jwang98052 replied:
Good point!

markusweimer commented on lang/cs/Org.Apache.REEF.Common/Tasks/Defaults/DefaultCloseTaskHandler.cs:
```diff
@@ -0,0 +1,50 @@
+            throw new NotImplementedException();
+        }
+
+        public void OnNext(ICloseEvent value)
```
This is dangerous. If a Driver sends `close` to the Task and the task doesn't implement that feature, it will never exit this way. 

The equivalent Java class `DefaultCloseHandler` instead fails the Task:

```java
  public void onNext(final CloseEvent closeEvent) {
    throw new RuntimeException("No EventHandler<CloseEvent> registered. Event received: " + closeEvent);
  }
```

This is the safer alternative.

jwang98052 replied:
That's right. Added.

markusweimer commented on lang/cs/Org.Apache.REEF.Common/Tasks/Exceptions/TaskCloseHandlerException.cs:
```diff
@@ -0,0 +1,43 @@
+    /// An exception that is thrown when the task close event
+    /// handler is not bound.
+    /// </summary>
+    internal sealed class TaskCloseHandlerException : Exception
```
Rename to `TaskCloseHandlerNotBoundException` to be more explicit about its role.

jwang98052 replied:
Done

markusweimer commented on lang/cs/Org.Apache.REEF.Common/Tasks/Exceptions/TaskCloseHandlerException.cs:
```diff
@@ -0,0 +1,43 @@
+    /// </summary>
+    internal sealed class TaskCloseHandlerException : Exception
+    {
+        public TaskCloseHandlerException(string message)
```
Do these constructors have to be `public`?

jwang98052 replied:
It can be internal.

markusweimer commented:
I'm done with a pass. We should also add a test that makes sure things fail correctly when no task stop handler is bound.

jwang98052 commented:
@markusweimer I have pushed the code that addressed the review comments.

markusweimer commented on lang/cs/Org.Apache.REEF.Common/Runtime/Evaluator/Task/TaskRuntime.cs:
```diff
@@ -220,6 +222,15 @@ public void Deliver(byte[] message)
+            catch (Exception ex)
+            {
+                var closeEx = new TaskCloseHandlerNotBoundException("Unable to close task.", ex);
+                Utilities.Diagnostics.Exceptions.CaughtAndThrow(closeEx, Level.Error, Logger);
```
This exception shouldn't be related to `TaskCloseHandlerNotBoundException`: You bind a default handler, so there is no way to have a situation where a handler isn't bound. In this case, the handler was in fact bound, but threw an exception in its own right. This should result in an `IFailedTask` on the Driver side which contains enough information such that the Driver knows that its call to `IRunningTask.Dispose()` initiated this failure.

jwang98052 replied:
That's right. I think I just not catch any exception here. If the handler for the task throw any exception, that would automatically go to Task fail case and the task exception will be passed back to the driver. 
I will add a test case for it.

markusweimer commented on lang/cs/Org.Apache.REEF.Common/Tasks/Defaults/DefaultTaskCloseHandler.cs:
```diff
@@ -0,0 +1,54 @@
+
+        public void OnNext(ICloseEvent value)
+        {
+            Utilities.Diagnostics.Exceptions.Throw(new Exception("No EventHandler<CloseEvent> registered. Event received: " + value), Logger);
```
We need to harmonize exception handling. In this class, 2 out of 3 times we use `throw new ...`, and 1 out of 3 times we use `Utilities.Diagnostics.Exceptions.Throw(new ...)`.

markusweimer replied:
Also, this is a case where the `TaskCloseHandlerNotBoundException` would be the appropriate exception type.

jwang98052 replied:
Done

markusweimer commented:
I've done another pass.

jwang98052 commented:
@markusweimer  I have addressed your comments and added the following test cases:
* Throw exception in close handler
* No task stop handler is bound.

jwang98052 commented:
Test case updated. Ready to for next review.

markusweimer commented:
LGTM. Will test.

markusweimer commented:
tests pass, will merge.
